pkg/apiserver/metrics/deployments.go

```diff
@@ -0,0 +1,168 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"k8s.io/kubernetes/pkg/api/rest"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"golang.org/x/net/context"
+)
+
+var (
+	variableLabels = []string{"name", "namespace"}
+
+	deploymentStatusObservedGenerationDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_observed_generation",
+		"The observed generation of this deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_replicas",
+		"The number of replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUpdatedReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_updated_replicas",
+		"The number of updated replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusAvailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_available_replicas",
+		"The number of available replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUnavailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_unavailable_replicas",
+		"The number of unavailable replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+)
+
+// DeploymentCollector collects various metrics of deployment objects.
+type DeploymentCollector struct {
+	Storage rest.StandardStorage
+}
+
+// Returns a new instance of DeploymentCollector.
+func NewDeploymentsCollector(s rest.StandardStorage) *DeploymentCollector {
+	return &DeploymentCollector{
+		Storage: s,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- deploymentStatusObservedGenerationDesc
+	ch <- deploymentStatusReplicasDesc
+	ch <- deploymentStatusUpdatedReplicasDesc
+	ch <- deploymentStatusAvailableReplicasDesc
+	ch <- deploymentStatusUnavailableReplicasDesc
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Collect(ch chan<- prometheus.Metric) {
+	var err error
+	deployments, err := c.Storage.List(context.Background(), nil)
+	if err != nil {
+		glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+	}
+
+	for _, d := range deployments.(*extensions.DeploymentList).Items {
+		err = addMetrics(ch, d)
+		if err != nil {
+			glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+		}
+	}
+}
+
+// addMetrics adds Metrics for a single deployment to the specified channel
+func addMetrics(ch chan<- prometheus.Metric, d extensions.Deployment) error {
```

Comment:
At 2-5k deployments, how would this impact a cluster?  Are all of these metrics observational (to state transitions)?  If so then I'd recommend we track this outside the apiservers as a first pass (from a design perspective and a system coupling perspective).

it would be best to add a small proposal to this PR that includes the trade offs between in apiserver and outside.